src/extras/cjs.js

```diff
@@ -0,0 +1,193 @@
+import babel from '@babel/core';
+import dynamicImportPlugin from '@babel/plugin-syntax-dynamic-import';
+import esRegisterFormatPlugin from '@babel/plugin-transform-modules-systemjs';
+
+import extractCjsDepsPlugin from '../utils/babel-visit-cjs-deps-plugin';
+import { compileScript } from '../utils/compile';
+import { sourceMapSources, basename, dirname } from '../common';
+
+
+async function createRequire(loader, url, deps, resolves) {
+  const mapImportSpecifierToUrl = async (specifier) => {
+    const resolvedUrl = await loader.resolve(specifier, url);
+    await loader.import(specifier, url);
+    return [specifier, resolvedUrl, true];
+  };
+
+  const mapResolveSpecifierToUrl = async (specifier) => {
+    const resolvedUrl = await loader.resolve(specifier, url);
+    return [specifier, resolvedUrl, false];
+  };
+
+  const importSpecifierMapEntries = await Promise.all(deps.map(mapImportSpecifierToUrl));
+  const resolveSpecifierMapEntries = await Promise.all(resolves.map(mapResolveSpecifierToUrl));
+  const specifierMapEntries = [...importSpecifierMapEntries, ...resolveSpecifierMapEntries];
+  const specifierMap = specifierMapEntries.reduce((acc, [specifier, resolvedUrl, required]) => {
+    const [, origRequired = false] = acc[specifier] || [];
+    acc[specifier] = [resolvedUrl, origRequired || required];
+    return acc;
+  }, {});
+
+  function require(specifier) {
+    const [resolvedUrl, required] = specifierMap[specifier];
+    if (!required) {
+      throw new Error(`Cannot require dynamic module '${specifier}'`);
+    }
+    return loader.get(resolvedUrl).default;
+  }
+
+  require.resolve = function resolve(specifier) {
+    const [resolvedUrl] = specifierMap[specifier];
+    return resolvedUrl;
+  };
+
+  return require;
+}
+
+function createModule(require, url) {
+  return {
+    children: [],
+    exports: Object.create(null),
+    filename: url.href,
+    id: url.href,
+    loaded: false,
+    parent: null,
+    paths: [],
+    require,
+  };
+}
+
+function postExecute(updateExport, module) {
+  const exports = module.exports;
+
+  if ((typeof exports === 'object' || typeof exports === 'function') && !('__esModule' in exports)) {
+    Object.defineProperty(module.exports, '__esModule', {
+      value: true,
+    });
+  }
+
+  updateExport('default', module.exports);
+
+  module.loaded = true;
+}
+
+
+function babelTransform(source, options) {
+  return new Promise(function(resolve, reject) {
+    babel.transform(source, options, function(err, result) {
```

Comment:
These seems like a separate concern to supporting CommonJS to me, that should be its own translator if desired.

Reply:
I would be fine turning this into a separate PR. Perhaps, it would be to create a babel transform that has uses `context` and supports sourcemap generation. It would also accept configuration/plugins (in the `context`, perhaps?) that it merges into its own configuration.

Is that what you were thinking?